llarp/service/outbound_context.hpp

```diff
@@ -10,29 +11,82 @@
 namespace llarp::service
 {
   struct AsyncKeyExchange;
+
   struct Endpoint;
 
   /// context needed to initiate an outbound hidden service session
-  struct OutboundContext : public path::Builder,
-                           public SendContext,
+  struct OutboundContext : public llarp::path::Builder,
                            public std::enable_shared_from_this<OutboundContext>
   {
+   private:
+    Endpoint& ep;
+
+    IntroSet current_intro;
+    Introduction next_intro;
+
+    const dht::Key_t location;
+    const Address addr;
+
+    ServiceInfo remote_identity;
+    Introduction remote_intro;
+
+    ConvoTag current_tag;
+
+    uint64_t update_introset_tx = 0;
+    uint16_t lookup_fails = 0;
+    uint16_t build_fails = 0;
+
+    bool got_inbound_traffic = false;
+    bool generated_convo_intro = false;
+    bool sent_convo_intro = false;
+    bool marked_bad = false;
+
+    const std::chrono::milliseconds created_at;
+    std::chrono::milliseconds last_send = 0ms;
+    std::chrono::milliseconds send_timeout = path::BUILD_TIMEOUT;
+    std::chrono::milliseconds connect_timeout = send_timeout * 2;
+    std::chrono::milliseconds last_shift = 0ms;
+    std::chrono::milliseconds last_inbound_traffic = 0ms;
+    std::chrono::milliseconds last_introset_update = 0ms;
+    std::chrono::milliseconds last_keep_alive = 0ms;
+
+    void
+    gen_intro_async_impl(
+        std::string payload, std::function<void(std::string, bool)> func = nullptr);
+
+   public:
     OutboundContext(const IntroSet& introSet, Endpoint* parent);
 
     ~OutboundContext() override;
 
+    ConvoTag
+    get_current_tag() const
+    {
+      return current_tag;
+    }
+
+    void
+    gen_intro_async(std::string payload);
+
+    void
+    encrypt_and_send(std::string buf);
+
+    /// for exits
+    void
+    send_packet_to_remote(std::string buf) override;
+
+    void
+    send_auth_async(std::function<void(std::string, bool)> resultHandler);
+
     void
-    Tick(llarp_time_t now) override;
+    Tick(std::chrono::milliseconds now) override;
```

Comment:
This (any many other in this file) should remain as `llarp_time_t` as they are representing timepoints (and even though `llarp_time_t` is a typedef to `ms`, it's useful also signals that a value is a time point, rather than a duration.  `llarp_time_t` isn't well implemented in that it really should be a proper time point, but that's a refactor for another day, and switching it to a plain duration is a step in the wrong direction here.

Looking at some of the others, it's possible that one or two of them should be converted when they are conceptually durations (and not conceptually time points), such as RTT().